From 8cd9ea9f32fc532163f737c43b9725ea72da5998 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Fri, 9 Oct 2026 17:37:37 +0800
Subject: [PATCH] GLInject: add support for MIPS, LoongArch, and RISC-V

Add R_JUMP_SLOT and ABI-specific R_GLOBAL_DATA definitions for MIPS,
LoongArch (LP64 and ILP32), and RISC-V (32 and 64-bit).

Please note that R_GLOBAL_DATA definitions were not done for MIPS, as
doing so involves grappling with ABI complexity, which I am not yet ready
for. Leave it as TODO here since it's optional.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 glinject/plthook_elf.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/glinject/plthook_elf.c b/glinject/plthook_elf.c
index a249cfd..a59db50 100644
--- a/glinject/plthook_elf.c
+++ b/glinject/plthook_elf.c
@@ -101,6 +101,27 @@
 #elif defined __powerpc__
 #define R_JUMP_SLOT   R_PPC_JMP_SLOT
 #define R_GLOBAL_DATA R_PPC_GLOB_DAT
+#elif defined __loongarch__
+#define R_JUMP_SLOT   R_LARCH_JUMP_SLOT
+#if defined __loongarch_lp64
+#define R_GLOBAL_DATA R_LARCH_64
+#elif defined __loongarch_ilp32
+#define R_GLOBAL_DATA R_LARCH_32
+#else
+#error Unknown LoongArch ABI
+#endif /* if defined __loongarch_lp64 */
+#elif defined __mips__
+#define R_JUMP_SLOT   R_MIPS_JUMP_SLOT
+/* TODO: R_GLOBAL_DATA not defined yet. */
+#elif defined __riscv
+#define R_JUMP_SLOT   R_RISCV_JUMP_SLOT
+#if __riscv_xlen == 32
+#define R_GLOBAL_DATA R_RISCV_32
+#elif __riscv_xlen == 64
+#define R_GLOBAL_DATA R_RISCV_64
+#else
+#error Unknown RISC-V ABI
+#endif /* if __riscv_xlen == 32 */
 #elif 0 /* disabled because not tested */ && (defined __sparcv9 || defined __sparc_v9__)
 #define R_JUMP_SLOT   R_SPARC_JMP_SLOT
 #elif 0 /* disabled because not tested */ && (defined __sparc || defined __sparc__)
-- 
2.56.0

